samples/will_it_shuffle.lobster

```diff
@@ -0,0 +1,228 @@
+// adaptation of https://bost.ocks.org/mike/shuffle/compare.html
+//
+// TODO
+// - vectorize the b-spline
+// - more shufflers
+// - measure reshuffle time
+// - swappable RNG
+
+import color
+import imgui
+import std
+import vec
+
+enum shuffle_mode:
+    shuffle_none = 0
+    shuffle_naive_swap
+    shuffle_push_remove
+    shuffle_fisher_yates
+
+var mode = shuffle_none
+
+def shuffle(array): switch mode:
+    case shuffle_none: array
+    case shuffle_naive_swap: naive_swap(array)
+    case shuffle_push_remove: push_remove(array)
+    case shuffle_fisher_yates: fisher_yates(array)
+
+def mode_name(): return switch mode:
+    case shuffle_none: "none"
+    case shuffle_naive_swap: "naive swap"
+    case shuffle_push_remove: "push remove"
+    case shuffle_fisher_yates: "fisher yates"
+    default: ""
+
+def naive_swap(xs):
+    // ala https://blog.codinghorror.com/the-danger-of-naivete/
+    for(xs) x, i:
+        let j = rnd(xs.length)
+        xs[i] = xs[j]
+        xs[j] = x
+
+def push_remove(xs):
+    // std.lobster original
+    for(xs.length) i:
+        xs.push(xs.remove(rnd(xs.length - i)))
+
+def fisher_yates(xs):
+    for(xs) x, i:
+        if i:
+            let j = rnd(i+1)
+            if j != i:
+                // FIXME want to write: xs[i], xs[j] = xs[j], xs[i]
+                let tmp = xs[i]
+                xs[i] = xs[j]
+                xs[j] = tmp
+
+// TODO more from https://bost.ocks.org/mike/shuffle/compare.html
+
+let colors = map([
+    // from https://github.com/d3/d3-scale-chromatic/blob/master/src/diverging/PuOr.js
+    0x2D004B,
+    0x542788,
+    0x8073AC,
+    0xB2ABD2,
+    0xD8DAEB,
+    0xF7F7F7,
+    0xFEE0B6,
+    0xFDB863,
+    0xE08214,
+    0xB35806,
+    0x7F3B08,
+]): color {
+    div( (_ >> 16) & 0xFF, 255),
+    div( (_ >> 8)  & 0xFF, 255),
+    div(  _        & 0xFF, 255),
+    1,
+}
+
+def spline_value(values, t):
+    // uniforem b-spline interpolation adapted from d3-scale-chromactic
+
+    // TODO can we translate from this Basis Spline (B-spline) to a cardinal
+    // spline? Or should lobster also provide a builtin for B-splines?
```

Comment:
Yup, feel free to add other spline functions, similar to the code for cardinal spline. That is nice to have, and generally good to have very re-usable core code in C++.

Reply:
Neat, will look into doing so after this PR; I think I'm going to actually try writing something more like a game first fwiw :-)